R/calcOutput.R

```diff
@@ -338,19 +338,57 @@ calcOutput <- function(type, aggregate = TRUE, file = NULL, years = NULL, round
   date        <- .prepComment(date(), "creation date")
   note        <- .prepComment(x$note, "note")
 
-  # select fitting relation mapping
+  # select fitting relation mappings and merge them if there is more than one
   if (aggregate != FALSE) {
     if (x$class != "magpie") stop("Aggregation can only be used in combination with x$class=\"magpie\"!")
     items <- getItems(x$x, dim = 1)
-    relFitting <- which(sapply(rel, nrow) == length(items)) # nolint
+
+    # find mappings that have the same (setequal) items (usually ISO countries) in any column as data (items)
+    # lapply loops over the mappings, vapply loops over the columns of a mapping
+    columnHasItems <- lapply(rel, vapply, setequal, items, FUN.VALUE = logical(1))
+    # extract the names of columns that have the same items as data
+    columnNameWithItems <- lapply(columnHasItems, function(x) names(which(x)))
+
+    # mappings must have the same length AND the same items as data
+    relFitting <- which(vapply(rel, nrow, FUN.VALUE = integer(1)) == length(items) &
+                        !vapply(columnNameWithItems, identical, character(0), FUN.VALUE = logical(1)))
+
     if (length(relFitting) == 0) stop("Neither getConfig(\"regionmapping\") nor getConfig(\"extramappings\")",
       " contain a mapping compatible to the provided data!")
+
+    # inform about mappings that don't fit and will be omitted
+    omit <- setdiff(names(rel), names(relFitting))
+    if (length(omit) > 0)
+      vcat(verbosity = 2, "Ignoring region mapping ", omit, " because it does not fit the data")
+
+    # keep mappings only that fit the data
+    rel <- rel[relFitting]
+
+    # if there is only one fitting mapping make rel a data frame
+    if (length(rel) == 1)
+      rel <- rel[[1]]
```

Comment:
same here, use curly braces to make it more visible what belongs into the if-statement